File: go/engine/conformance/runner.go
//go:build selfhost

package conformance

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	engine "github.com/ball-lang/ball/go/engine"
)

// perFixtureTimeout bounds each fixture so a latent infinite loop cannot wedge
// the whole sweep (matches the Rust/C# runners' 120 s budget). Override with
// BALL_TIMEOUT_MS for faster iteration sweeps. This is the *cooperative* budget:
// runOne feeds it to the compiled engine, whose per-expression timeout guard
// makes a runaway self-abort at this point (issue #436) so its goroutine exits
// rather than spinning for the rest of the sweep.
func perFixtureTimeout() time.Duration {
	if ms := os.Getenv("BALL_TIMEOUT_MS"); ms != "" {
		if n, err := strconv.Atoi(ms); err == nil && n > 0 {
			return time.Duration(n) * time.Millisecond
		}
	}
	return 120 * time.Second
}

// hardDeadlineGrace is how much longer runOne waits past the cooperative budget
// before falling back to the select backstop. In the flat-stack runaway case
// (an infinite while/for) the engine self-aborts at the budget, delivers its
// timeout on the buffered channel, and the goroutine exits — well before this
// grace elapses. The backstop fires for the runaways the cooperative guard does
// NOT reliably stop: (a) a native loop inside a runtime helper, which never
// returns to an expression eval, and (b) unbounded-stack Ball RECURSION, whose
// per-level guard checks were observed not to abort within the budget. In both
// cases Go cannot kill the goroutine in-process, so it keeps spinning (leaks)
// after the backstop reports the timeout — and unbounded recursion additionally
// risks a fatal Go stack overflow under the driver's 1 GiB stack ceiling, which
// would kill the whole sweep binary. Only flat-stack runaways are truly
// self-aborting; the backstop keeps the sweep moving for the rest.
const hardDeadlineGrace = 10 * time.Second

// isExecutionTimeout reports whether err is the compiled engine's cooperative
// execution-timeout self-abort (BallRuntimeError('Execution timeout exceeded'),
// surfaced through the driver's panic recovery). Such a fixture is reported as a
// "timeout", not a generic "error".
func isExecutionTimeout(err error) bool {
	return err != nil && strings.Contains(err.Error(), "Execution timeout exceeded")
}

// Result is one fixture's outcome.
type Result struct {
	Name   string
	Status string // "pass", "fail", "timeout", "error"
	Detail string
}

// Summary is a whole-corpus sweep outcome.
type Summary struct {
	Passed  int
	Failed  int
	Total   int
	Skipped int // golden-less carve-outs
	Results []Result
}

// RunAll drives every tests/conformance/*.ball.json fixture through the compiled
// self-hosted engine and compares stdout to its .expected_output.txt golden. A
// fixture with no golden is a documented carve-out and is skipped (like the Dart
// runner). If onlyFixture is non-empty, only that fixture runs.
func RunAll(onlyFixture string) (Summary, error) {
	dir, err := conformanceDir()
	if err != nil {
		return Summary{}, err
	}
	entries, err := filepath.Glob(filepath.Join(dir, "*.ball.json"))
	if err != nil {
		return Summary{}, err
	}
	sort.Strings(entries)

	var s Summary
	for _, path := range entries {
		name := strings.TrimSuffix(filepath.Base(path), ".ball.json")
		if onlyFixture != "" && name != onlyFixture {
			continue
		}
		goldenPath := strings.TrimSuffix(path, ".ball.json") + ".expected_output.txt"
		golden, gerr := os.ReadFile(goldenPath)
		if gerr != nil {
			s.Skipped++
			continue
		}
		r := runOne(name, path, string(golden))
		s.Results = append(s.Results, r)
		s.Total++
		if r.Status == "pass" {
			s.Passed++
		} else {
			s.Failed++
		}
	}
	return s, nil
}

func runOne(name, path, golden string) Result {
	data, err := os.ReadFile(path)
	if err != nil {
		return Result{name, "error", err.Error()}
	}

	budget := perFixtureTimeout()

	type outcome struct {
		out []string
		err error
	}
	ch := make(chan outcome, 1)
	go func() {
		defer func() {
			if r := recover(); r != nil {
				ch <- outcome{nil, asError(r)}
			}
		}()
		eng, err := engine.FromJSON(data)
		if err != nil {
			ch <- outcome{nil, err}
			return
		}
		// Drive the compiled engine's cooperative execution-timeout guard so a
		// runaway fixture self-aborts (and this goroutine then exits) instead of
		// spinning for the rest of the sweep — Go cannot kill a goroutine, so the
		// select backstop below alone would leak it (issue #436).
		eng.TimeoutMs = budget.Milliseconds()
		out, err := eng.Run()
		ch <- outcome{out, err}
	}()

	select {
	case res := <-ch:
		if res.err != nil {
			// A cooperative self-abort at the budget reports as a timeout, not a
			// generic error; the goroutine has already exited by the time we get
			// here (it delivered on the buffered channel).
			if isExecutionTimeout(res.err) {
				return Result{name, "timeout", ""}
			}
			return Result{name, "error", firstLine(res.err.Error())}
		}
		actual := strings.TrimRight(strings.Join(res.out, "\n"), "\n\r")
		expected := strings.TrimRight(strings.ReplaceAll(golden, "\r\n", "\n"), "\n")
		if actual == expected {
			return Result{name, "pass", ""}
		}
		return Result{name, "fail", diffDetail(expected, actual)}
	case <-time.After(budget + hardDeadlineGrace):
		// Backstop: the cooperative guard did not abort in time. This happens for
		// a native loop inside a runtime helper (never returns to an expression
		// eval) AND for unbounded-stack Ball recursion (guard checks run but were
		// observed not to abort within the budget). Report the timeout and move
		// on — but this goroutine is still running and LEAKS (Go cannot kill it),
		// and a recursing one may yet fatal the sweep via Go stack overflow. Only
		// flat-stack runaways (while/for) are reliably self-aborted upstream.
		return Result{name, "timeout", "cooperative timeout not observed"}
	}
}

func diffDetail(expected, actual string) string {
	el := strings.Split(expected, "\n")
	al := strings.Split(actual, "\n")
	if os.Getenv("BALL_FIXTURE") != "" {
		return "\n--- expected (" + strconv.Itoa(len(el)) + ") ---\n" + expected +
			"\n--- actual (" + strconv.Itoa(len(al)) + ") ---\n" + actual
	}
	return "expected(" + strconv.Itoa(len(el)) + "): " + first(el) + " | actual(" + strconv.Itoa(len(al)) + "): " + first(al)
}

func first(xs []string) string {
	if len(xs) == 0 {
		return "<none>"
	}
	return xs[0]
}

func firstLine(s string) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		return s[:i]
	}
	return s
}

func asError(r any) error {
	if e, ok := r.(error); ok {
		return e
	}
	return &recoveredError{r}
}

type recoveredError struct{ v any }

func (e *recoveredError) Error() string { return fmt.Sprint(e.v) }

// conformanceDir walks up from the test's working directory to the repo root and
// returns tests/conformance.
func conformanceDir() (string, error) {
	dir, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		candidate := filepath.Join(dir, "tests", "conformance")
		if fi, err := os.Stat(candidate); err == nil && fi.IsDir() {
			return candidate, nil
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", os.ErrNotExist
		}
		dir = parent
	}
}

File: python/engine/conformance/runner.py
"""Whole-corpus conformance sweep for the self-hosted Python engine.

Drives every ``tests/conformance/*.ball.json`` fixture through the compiled
engine and compares stdout to its ``.expected_output.txt`` golden, printing the
CI-parseable::

    Results: N passed, M failed, T total (K skipped carve-outs)

plus a ``FAILING [name] status detail`` line per non-passing fixture.

Each fixture runs as its own ``python -m ball_engine`` **subprocess** with a
per-fixture timeout, so a runaway (infinite loop / unbounded recursion) is simply
killed — a Python process is trivially killable, which sidesteps the goroutine
-leak problem the Go runner has to work around cooperatively. Fixtures run in
parallel (subprocess.run releases the GIL while waiting) to keep the ~320-fixture
sweep well under the wall-clock budget.

A fixture with no golden is a documented carve-out and is skipped (the same 4
resource-limit/sandbox fixtures Rust/C#/Go skip). ``BALL_FIXTURE=<name>`` runs a
single fixture with a full diff.
"""

from __future__ import annotations

import concurrent.futures
import glob
import os
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path

CARVEOUTS = {"196_timeout", "197_memory_limit", "201_input_validation", "202_sandbox_mode"}

_HERE = Path(__file__).resolve()
_ENGINE_DIR = _HERE.parents[1]                 # python/engine
_REPO_ROOT = _HERE.parents[3]                  # repo root
_CONFORMANCE = _REPO_ROOT / "tests" / "conformance"

# Per-fixture wall-clock budget (subprocess kill). Overridable for fast sweeps.
_TIMEOUT_S = float(os.environ.get("BALL_TIMEOUT_S", "120"))
# Cooperative execution-timeout the engine self-aborts at, a hair under the kill.
_ENGINE_TIMEOUT_MS = int(_TIMEOUT_S * 1000) - 5000
_WORKERS = int(os.environ.get("BALL_WORKERS", str(min(8, (os.cpu_count() or 4)))))


@dataclass
class Result:
    name: str
    status: str   # pass | fail | timeout | error
    detail: str = ""


@dataclass
class Summary:
    passed: int = 0
    failed: int = 0
    total: int = 0
    skipped: int = 0
    results: list = None


def _run_one(name: str, path: str, golden: str) -> Result:
    env = dict(os.environ)
    env["BALL_TIMEOUT_MS"] = str(max(1000, _ENGINE_TIMEOUT_MS))
    try:
        proc = subprocess.run(
            [sys.executable, "-m", "ball_engine", path],
            cwd=str(_ENGINE_DIR),
            env=env,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=_TIMEOUT_S,
        )
    except subprocess.TimeoutExpired:
        return Result(name, "timeout", f"killed after {_TIMEOUT_S:.0f}s")

    actual = proc.stdout.replace("\r\n", "\n").rstrip("\n")
    expected = golden.replace("\r\n", "\n").rstrip("\n")
    if actual == expected:
        return Result(name, "pass")
    if proc.returncode != 0:
        err = (proc.stderr or "").strip().splitlines()
        detail = err[-1] if err else "non-zero exit"
        if "Execution timeout exceeded" in (proc.stderr or ""):
            return Result(name, "timeout", "execution timeout")
        return Result(name, "error", detail[:200])
    el = expected.split("\n")
    al = actual.split("\n")
    if os.environ.get("BALL_FIXTURE"):
        detail = (f"\n--- expected ({len(el)}) ---\n{expected}"
                  f"\n--- actual ({len(al)}) ---\n{actual}")
    else:
        exp0 = el[0] if el else "<none>"
        act0 = al[0] if al else "<none>"
        detail = f"expected({len(el)}): {exp0} | actual({len(al)}): {act0}"
    return Result(name, "fail", detail)


def run_all(only: str = "") -> Summary:
    paths = sorted(glob.glob(str(_CONFORMANCE / "*.ball.json")))
    jobs = []
    skipped = 0
    for path in paths:
        name = Path(path).name[: -len(".ball.json")]
        if only and name != only:
            continue
        golden_path = path[: -len(".ball.json")] + ".expected_output.txt"
        if not os.path.exists(golden_path):
            skipped += 1
            continue
        golden = Path(golden_path).read_text(encoding="utf-8")
        jobs.append((name, path, golden))

    results: list[Result] = []
    with concurrent.futures.ThreadPoolExecutor(max_workers=_WORKERS) as ex:
        futs = {ex.submit(_run_one, n, p, g): n for (n, p, g) in jobs}
        for fut in concurrent.futures.as_completed(futs):
            results.append(fut.result())

    results.sort(key=lambda r: r.name)
    s = Summary(results=results, skipped=skipped)
    for r in results:
        s.total += 1
        if r.status == "pass":
            s.passed += 1
        else:
            s.failed += 1
    return s


def main() -> int:
    only = os.environ.get("BALL_FIXTURE", "")
    s = run_all(only)
    for r in s.results:
        if r.status != "pass":
            print(f"FAILING [{r.name}] {r.status} {r.detail}")
    print(f"Results: {s.passed} passed, {s.failed} failed, {s.total} total "
          f"({s.skipped} skipped carve-outs)")
    return 0 if s.failed == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())
